### PR TITLE
Add `array_set` scalar function

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
This scalar function is a prerequisite for supporting updating array by the elements, see https://github.com/crate/crate/pull/14140#pullrequestreview-1450330940.

NOTE: 
1) negative indexes are rejected currently.
2) there is no circuit breaker in place for a large update resulting in out of memory error - ex) `update t set a[2147483647] = 1` (which adds a huge null padding). 

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: libs/shared/src/main/java/io/crate/common/collections/Lists2.java [hunk 1]
```diff
@@ -101,7 +101,7 @@ public static boolean equals(Collection<?> list1, Collection<?> list2) {
      */
     public static <I, O> List<O> map(Collection<I> list, Function<? super I, ? extends O> mapper) {
         if (list.isEmpty()) {
-            return List.of();
+            return new ArrayList<>();
```
Comment (jeeminso) [inline libs/shared/src/main/java/io/crate/common/collections/Lists2.java:104]: This allows empty arrays to be mutable such that values can be added.

Comment (mkleen) [reply libs/shared/src/main/java/io/crate/common/collections/Lists2.java:104]: We try to keep all collections by default immutable. There are exceptions to this for performance reason, but I think the core libraries should follow this rule. Where is the code part where you need to add to the list after a map operation? Can you not just create a new mutable List there ?

Comment (matriv) [reply libs/shared/src/main/java/io/crate/common/collections/Lists2.java:104]: I agree, we should't do this here in `Lists2` or if it turns out to be the only convenient way, we should at least have a separate method explaining this behaviour (mutable List).

**Review by matriv — COMMENTED** [review]
thx @jeeminso ! Left an idea on how to avoid creating mutable Lists before hand.

File: libs/shared/src/main/java/io/crate/common/collections/Lists2.java (hunk 1 above)
Comment (jeeminso) [reply libs/shared/src/main/java/io/crate/common/collections/Lists2.java:104]: Sorry this was my mistake. While extracting `array_set` from my other PR, I carried this over by accident. Reverting these `new ArrayList<>()` should not break anything. I will keep this in mind for the remaining parts of #14140.

Comment (mkleen) [reply libs/shared/src/main/java/io/crate/common/collections/Lists2.java:104]: 👍

File: server/src/main/java/io/crate/expression/scalar/ArraySetFunction.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        }
+        return updated;
+    }
+
+    private void setElement(List<Object> source, int index, Object value) {
+        if (index <= 0) {
+            throw new UnsupportedOperationException("Updating arrays with indexes <= 0 is not supported");
+        }
+        index--; // Since array indexes in CrateDB start from 1
+        if (index < source.size()) {
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/expression/scalar/ArraySetFunction.java:105]: Sorry please check my other comment, https://github.com/crate/crate/pull/14229/files#r1221198052.

**Review by mkleen — APPROVED** [review]
Looks good to me! Great work!

**Review by matriv — APPROVED** [review]
LGTM, thx! Left a comment to clarify in the docs, that `0` index is also invalid.

**Comment by jeeminso** [comment]
retest this please

**Pull request merged**